src/query.js

```diff
@@ -58,266 +64,565 @@ class Query {
+   *
+   * @returns {Object}
+   */
+  setupQueue () {
+    const q = queue(this.processNext.bind(this), this.concurrency)
+
+    // If there's an error, stop the worker
+    q.error = (err) => {
+      this.log.error('queue', err)
+      this.stop(err)
+    }
+
+    // When all peers in the queue have been processed, stop the worker
+    q.drain = () => {
+      this.log('queue:drain')
+      this.stop()
+    }
+
+    // When a space opens up in the queue, add some more peers
+    q.unsaturated = () => {
+      if (this.running) {
+        this.log('queue:unsaturated')
+        this.fill()
+      }
+    }
+
+    q.buffer = 0
+
+    return q
+  }
+
+  /**
+   * Stop the worker, optionally providing an error to pass to the worker's
+   * callback.
+   *
+   * @param {Error} err
+   */
+  stop (err) {
+    if (!this.running) {
+      return
+    }
+
+    this.running = false
+    this.queue.kill()
+    this.callbackFn(err)
   }
 
   /**
    * Use the queue from async to keep `concurrency` amount items running
    * per path.
    *
-   * @param {Query} query
-   * @param {Object} path
    * @param {function(Error)} callback
    */
-  workerQueue (query, path, callback) {
-    let workerRunning = true
-    const q = queue((next, cb) => {
-      query._log('queue:work')
-      this.execQuery(next, query, path, (err, state) => {
+  execute (callback) {
+    this.running = true
+    this.callbackFn = callback
+    this.fill()
+  }
+
+  /**
+   * Add peers to the worker queue until there are enough to satisfy the
+   * worker queue concurrency.
+   * Note that we don't want to take any more than those required to satisfy
+   * concurrency from the peers-to-query queue, because we always want to
+   * query the closest peers to the key first, and new peers are continously
+   * being added to the peers-to-query queue.
+   */
+  fill () {
+    this.log('queue:fill')
+
+    // Note:
+    // - queue.running(): number of items that are currently running
+    // - queue.length(): the number of items that are waiting to be run
+    while (this.queue.running() + this.queue.length() < this.concurrency &&
+           this.path.peersToQuery.length > 0) {
+      this.queue.push(this.path.peersToQuery.dequeue())
+    }
+  }
+
+  /**
+   * Process the next peer in the queue
+   *
+   * @param {PeerId} peer
+   * @param {function(Error)} cb
+   * @returns {void}
+   */
+  processNext (peer, cb) {
+    if (!this.running) {
+      return
+    }
+
+    // The paths must be disjoint, meaning that no two paths in the Query may
+    // traverse the same peer
+    if (this.run.peersSeen.has(peer)) {
+      return cb()
+    }
+
+    // Check if we've queried enough peers already
+    this.run.continueQuerying((err, continueQuerying) => {
+      if (!this.running) {
+        return
+      }
+
+      if (err) {
+        return cb(err)
+      }
+
+      // If we've queried enough peers, bail out
+      if (!continueQuerying) {
+        return
```

Comment:
In a few places in this function, such as here, it's possible that `cb` never gets called. Is this intended? Will this potentially lead to stalled/stuck queues?


Other than this, things look good!

Reply:
It only ever returns without calling `cb` after the workers have all been stopped, but I think it's less confusing if we call `cb` in that situation anyway, so I changed it to do so.